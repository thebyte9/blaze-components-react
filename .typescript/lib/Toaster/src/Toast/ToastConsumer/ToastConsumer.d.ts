/// <reference types="react" />
import { IContext } from "../types/provider/provider";
declare const ToastConsumer: ({ children }: {
    children: (context: IContext) => JSX.Element;
}) => JSX.Element;
export default ToastConsumer;
