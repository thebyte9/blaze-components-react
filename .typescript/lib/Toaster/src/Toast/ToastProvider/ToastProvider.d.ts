import { Component } from "react";
import { Callback, IOptions } from "../types/common";
import { IProps, IState } from "../types/provider/provider";
declare class ToastProvider extends Component<IProps, IState> {
    static defaultProps: {
        autoDismiss: boolean;
        autoDismissTimeout: number;
        components: {
            Toast: {
                ({ appearance, autoDismiss, autoDismissTimeout, children, isRunning, onDismiss, placement, transitionDuration, transitionState, onMouseEnter, onMouseLeave, ...otherProps }: import("../types/common").IToastProps): JSX.Element;
                defaultProps: {
                    onDismiss: () => void;
                };
            };
            ToastContainer: ({ hasToasts, placement, ...props }: any) => JSX.Element;
        };
        placement: string;
        transitionDuration: number;
    };
    state: {
        toasts: never[];
    };
    has: (id: string) => boolean;
    onDismiss: (id: string, cb?: Callback) => () => void;
    add: (content: JSX.Element, options?: IOptions, cb?: Callback) => any;
    remove: (id: string, cb?: Callback) => void;
    removeAll: () => void;
    update: (id: string, options?: IOptions, cb?: Callback) => void;
    render(): JSX.Element;
}
export default ToastProvider;
