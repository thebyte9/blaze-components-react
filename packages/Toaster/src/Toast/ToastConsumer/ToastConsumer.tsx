import React from "react";
import ToastContext from "../ToastContext/ToastContext";
import { IContext } from "../types/provider/provider";

const ToastConsumer = ({
  children
}: {
  children: (context: IContext) => JSX.Element;
}) => (
  <ToastContext.Consumer>
    {(context: any) => children(context)}
  </ToastContext.Consumer>
);

export default ToastConsumer;
