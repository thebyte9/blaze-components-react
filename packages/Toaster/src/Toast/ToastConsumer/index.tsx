import ToastConsumer from "./ToastConsumer";

export default ToastConsumer;
