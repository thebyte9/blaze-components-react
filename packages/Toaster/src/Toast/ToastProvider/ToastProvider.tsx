import React, { Component } from "react";
import { createPortal } from "react-dom";
import { Transition, TransitionGroup } from "react-transition-group";
import {
  DEFAULT_AUTO_DISMISS_TIMEOUT,
  DEFAULT_PLACEMENT,
  DEFAULT_TRANSITION_DURATION
} from "../constants";
import DefaultToast from "../DefaultToast";
import ToastContainer from "../ToastContainer/ToastContainer";
import ToastContext from "../ToastContext";
import ToastController from "../ToastController";
import { Callback, Id, IOptions } from "../types/common";
import { IProps, IState } from "../types/provider/provider";
import { generateUEID, NOOP } from "../utils";

const canUseDOM = !!(
  typeof window !== "undefined" &&
  window.document &&
  window.document.createElement
);

class ToastProvider extends Component<IProps, IState> {
  public static defaultProps = {
    autoDismiss: false,
    autoDismissTimeout: DEFAULT_AUTO_DISMISS_TIMEOUT,
    components: { Toast: DefaultToast, ToastContainer },
    placement: DEFAULT_PLACEMENT,
    transitionDuration: DEFAULT_TRANSITION_DURATION
  };

  public state = { toasts: [] };

  public has = (id: string) => {
    if (!this.state.toasts.length) {
      return false;
    }

    return Boolean(this.state.toasts.filter((t: any) => t.id === id).length);
  };
  public onDismiss = (id: Id, cb: Callback = NOOP) => () => {
    cb(id);
    this.remove(id);
  };

  public add = (
    content: JSX.Element,
    options: IOptions = {},
    cb: Callback = NOOP
  ) => {
    const id = options.id || generateUEID();
    const callback = () => cb(id);

    if (this.has(id)) {
      return;
    }

    this.setState((state: any): any => {
      const newToast = { content, id, ...options };
      const toasts = [...state.toasts, newToast];

      return { toasts };
    }, callback);

    return id;
  };
  public remove = (id: Id, cb: Callback = NOOP) => {
    const callback = () => cb(id);

    if (!this.has(id)) {
      return;
    }

    this.setState(state => {
      const toasts = state.toasts.filter(t => t.id !== id);
      return { toasts };
    }, callback);
  };
  public removeAll = () => {
    if (!this.state.toasts.length) {
      return;
    }

    this.state.toasts.forEach((t: any) => this.remove(t.id));
  };
  public update = (id: Id, options: IOptions = {}, cb: Callback = NOOP) => {
    const callback = () => cb(id);

    if (!this.has(id)) {
      return;
    }

    this.setState(state => {
      const old = state.toasts;
      const i = old.findIndex(t => t.id === id);
      const updatedToast = { ...old[i], ...options };
      const toasts = [...old.slice(0, i), updatedToast, ...old.slice(i + 1)];

      return { toasts };
    }, callback);
  };

  public render() {
    const {
      autoDismiss: inheritedAutoDismiss,
      autoDismissTimeout,
      children,
      components,
      placement,
      transitionDuration
    } = this.props;
    const { Toast, ToastContainer } = { ...components };
    const { add, remove, removeAll, update } = this;
    const toasts = Object.freeze(this.state.toasts);

    const hasToasts = Boolean(toasts.length);
    const portalTarget = canUseDOM ? document.body : null;

    return (
      <ToastContext.Provider value={{ add, remove, removeAll, update, toasts }}>
        {children}

        {portalTarget ? (
          createPortal(
            <ToastContainer placement={placement} hasToasts={hasToasts}>
              <TransitionGroup component={null}>
                {toasts.map(
                  ({
                    appearance,
                    autoDismiss,
                    content,
                    id,
                    onDismiss,
                    ...unknownConsumerProps
                  }: any) => (
                    <Transition
                      appear
                      key={id}
                      mountOnEnter
                      timeout={transitionDuration}
                      unmountOnExit
                    >
                      {transitionState => (
                        <ToastController
                          appearance={appearance}
                          autoDismiss={
                            autoDismiss !== undefined
                              ? autoDismiss
                              : inheritedAutoDismiss
                          }
                          autoDismissTimeout={autoDismissTimeout}
                          component={Toast}
                          key={id}
                          onDismiss={this.onDismiss(id, onDismiss)}
                          placement={placement}
                          transitionDuration={transitionDuration}
                          transitionState={transitionState}
                          {...unknownConsumerProps}
                        >
                          {content}
                        </ToastController>
                      )}
                    </Transition>
                  )
                )}
              </TransitionGroup>
            </ToastContainer>,
            portalTarget
          )
        ) : (
          <ToastContainer placement={placement} hasToasts={hasToasts} /> // keep ReactDOM.hydrate happy
        )}
      </ToastContext.Provider>
    );
  }
}

export default ToastProvider;
