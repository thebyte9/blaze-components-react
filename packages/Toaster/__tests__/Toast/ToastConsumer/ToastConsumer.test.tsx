import { render } from "@testing-library/react";
import expect from "expect";
import React from "react";
import ToastConsumer from "../../../src/Toast/ToastConsumer";
import ToastProvider from "../../../src/Toast/ToastProvider";

describe("Toast consumer", () => {
  it("should be defined", () => {
    expect(ToastConsumer).toBeDefined();
  });

  it("should pass the context to the children", () => {
    const FakeChildrenImplementation: any = jest.fn(() => (
      <div>Fake.ToastConsumer.Children.immplementation</div>
    ));

    const Provider: any = ToastProvider;

    const { container } = render(
      <Provider>
        <ToastConsumer>
          {ctx => <FakeChildrenImplementation {...ctx} />}
        </ToastConsumer>
      </Provider>
    );

    expect(container).toMatchSnapshot();
    expect(FakeChildrenImplementation).toHaveBeenCalledWith(
      expect.objectContaining({
        add: expect.any(Function),
        remove: expect.any(Function),
        removeAll: expect.any(Function),
        toasts: [],
        update: expect.any(Function)
      }),
      {}
    );
  });
});
