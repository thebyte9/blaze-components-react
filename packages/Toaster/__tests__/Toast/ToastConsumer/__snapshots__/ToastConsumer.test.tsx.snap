// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast consumer should pass the context to the children 1`] = `
<div>
  <div>
    Fake.ToastConsumer.Children.immplementation
  </div>
</div>
`;
